src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs

```diff
@@ -0,0 +1,313 @@
+
+namespace Nethermind.Synchronization.SnapSync;
+
+public class SnapServer : ISnapServer
+{
+    private readonly IReadOnlyTrieStore _store;
+    private readonly TrieStoreWithReadFlags _storeWithReadFlag;
+    private readonly IReadOnlyKeyValueStore _codeDb;
+    private readonly ILogManager _logManager;
+    private readonly ILogger _logger;
+
+    // On flatdb/halfpath, using ReadAhead flag significantly reduce IOPs by reading a larger chunk of sequential data.
+    // It also skip the block cache, which reduces impact on block processing.
+    // On hashdb, this causes each IOP to be significantly larger, so it make it a lot slower than it already is.
+    private readonly ReadFlags _optimizedReadFlags = ReadFlags.HintCacheMiss;
+
+    // Some cache to make sure we don't accidentally serve blocks with missing trie node.
+    // We don't know for sure if a tree was pruned with just the root hash.
+    private readonly LruCache<ValueHash256, DateTimeOffset> _rootWithMissingNode = new LruCache<ValueHash256, DateTimeOffset>(256, "");
+
+    private readonly AccountDecoder _decoder = new AccountDecoder();
+
+    private const long HardResponseByteLimit = 2000000;
+    private const int HardResponseNodeLimit = 100000;
+
+    public SnapServer(IReadOnlyTrieStore trieStore, IReadOnlyKeyValueStore codeDb, ILogManager logManager)
+    {
+        _store = trieStore ?? throw new ArgumentNullException(nameof(trieStore));
+        _storeWithReadFlag = new TrieStoreWithReadFlags(_store, _optimizedReadFlags);
+        _codeDb = codeDb ?? throw new ArgumentNullException(nameof(codeDb));
+        _logManager = logManager ?? throw new ArgumentNullException(nameof(logManager));
+        _logger = logManager.GetClassLogger();
+    }
+
+    private bool IsRootMissing(in ValueHash256 rootHash)
+    {
+        // TODO: Before checking missing node cache, actually check StateReader.
+        if (_rootWithMissingNode.TryGet(rootHash, out DateTimeOffset missingTime))
+        {
+            if ((DateTimeOffset.Now - missingTime) < TimeSpan.FromSeconds(10))
+            {
+                return true;
+            }
+
+            // So, problem is, it could be that the missing root was just not processed, so we can't just not retry
+            // forever. `DateTimeOffset.Now` is probably heavy though.
+            _rootWithMissingNode.Delete(rootHash);
+        }
+
+        return false;
+    }
+
+    private void TrackTrieNodeException(in ValueHash256 rootHash)
+    {
+        _rootWithMissingNode.Set(rootHash, DateTimeOffset.Now);
+    }
+
+    public byte[][]? GetTrieNodes(PathGroup[] pathSet, in ValueHash256 rootHash, CancellationToken cancellationToken)
+    {
+        if (IsRootMissing(rootHash)) return Array.Empty<byte[]>();
+
+        _logger.Debug($"Trie node {pathSet.Length}");
+        // TODO: use cache to reduce node retrieval from disk
+        int pathLength = pathSet.Length;
+        List<byte[]> response = new(pathLength);
```

Comment:
could we use ArrayPoolList, that would get disposed after the message is sent?

Reply:
used ArrayPool for local construction, but could we keep it until we send message and then dispose?